src/service.py

```diff
@@ -0,0 +1,165 @@
+            content: the data to be written to the file.
+            mode: access permission mask applied to the
+              file using chmod (e.g. 0o640).
+        """
+        with open(path, "w+", encoding="utf-8") as file:
+            file.write(content)
+        # Ensure correct permissions are set on the file.
+        os.chmod(path, mode)
+        try:
+            # Get the uid/gid for the root user (running the service).
+            # TODO: the user running the jenkins agent is currently root
+            # we should replace this by defining a dedicated user in the apt package
+            u = pwd.getpwnam("root")
+            # Set the correct ownership for the file.
+            os.chown(path, uid=u.pw_uid, gid=u.pw_gid)
+        except KeyError:
+            # Ignore non existing user error when it wasn't created yet.
+            pass
+
+    @property
+    def is_active(self) -> bool:
+        """Indicate if the jenkins agent service is active."""
+        return systemd.service_running(AGENT_SERVICE_NAME)
+
+    def install(self) -> None:
+        """Install and set up the jenkins agent apt package.
+
+        Raises:
+            ServiceRestartError: if the installation of the snap failed.
+            PackageInstallError: if the package installation failed.
+        """
+        try:
+            # Add ppa that hosts the jenkins-agent package
+            repositories = apt.RepositoryMapping()
+            if "deb-ppa.launchpadcontent.net/tphan025/ppa-jammy" not in repositories:
+                repositories.add(
+                    apt.DebianRepository(
+                        enabled=True,
+                        repotype="deb",
+                        uri="https://ppa.launchpadcontent.net/tphan025/ppa/ubuntu/",
+                        # TODO: depends on the series of the charm unit
+                        # set the release accordingly
+                        release="jammy",
+                        groups=["main"],
+                    )
+                )
+                apt.import_key(PPA_GPG_KEY_ID)
+            # Install the apt package
+            apt.update()
+            # Dependencies
+            apt.add_package("openjdk-17-jre")
+            # Jenkins agent package from PPA
+            apt.add_package(AGENT_PACKAGE_NAME)
+            systemd.service_restart(AGENT_SERVICE_NAME)
+        except systemd.SystemdError as exc:
+            raise ServiceRestartError("Error starting the agent service") from exc
+        except (apt.PackageError, apt.PackageNotFoundError) as exc:
+            raise PackageInstallError("Error installing the agent package") from exc
+
+    def restart(self) -> None:
+        """Start the agent service.
+
+        Raises:
+            ServiceRestartError: when restarting the service fails
+        """
+        # Render template and write to appropriate file if only credentials are set
+        if credentials := self.state.agent_relation_credentials:
```

Comment:
It is a bit confusing that it is possible to call this method with the credentials not present. I.e., the method is supposed to restart but it can't if the credentials are not there. Should missing credentials be an error case for this method?

Reply:
To clarify, this `if` statement only covers the configuration file rendering, and the restarting of the service happens a bit later in the method and is not in the scope of this `if` statement. If the credentials are not there for any reason, the service still gets restarted, but the configuration file is not written to the filesystem.

Reply:
So, in that case, will restarting the service result in the intention? I.e., can the service be started without the configuration file?

Reply:
That's a good point, as it is now the service will return an error if the configuration file is not there, and at this point of the workflow we should already established that the credentials have been received and are complete. I'll remove the conditional statement from this method.

Reply:
I also added error handling for cases where credentials are empty